{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "f0957533",
   "metadata": {},
   "outputs": [],
   "source": [
    "class Node:\n",
    "    \"\"\" Class Node to represent a vertex in the de bruijn graph \"\"\"\n",
    "    def __init__(self, lab):\n",
    "        self.label = lab\n",
    "        self.indegree = 0\n",
    "        self.outdegree = 0\n",
    "\n",
    "class Edge:\n",
    "    def __init__(self, lab):\n",
    "        self.label = lab\n",
    "\n",
    "def read_reads(fname):\n",
    "    \"\"\" Read short reads in FASTA format. It is assumed that one line in the input file correspond to one read. \"\"\"\n",
    "    f = open(fname, 'reads')\n",
    "    lines = f.readlines()\n",
    "    f.close()\n",
    "    reads = []\n",
    "\n",
    "    for line in lines:\n",
    "        if line[0] != '>':\n",
    "            reads = reads + [line.rstrip()]\n",
    "\n",
    "    return reads\n",
    "\n",
    "def construct_graph(reads, k):\n",
    "    \"\"\" Construct de bruijn graph from sets of short reads with k length word\"\"\"\n",
    "    edges = dict()\n",
    "    vertices = dict()\n",
    "\n",
    "    for read in reads:\n",
    "        i = 0\n",
    "        while i+k < len(read):\n",
    "            v1 = read[i:i+k]\n",
    "            v2 = read[i+1:i+k+1]\n",
    "            if v1 in edges.keys():\n",
    "                vertices[v1].outdegree += 1\n",
    "                edges[v1] += [Edge(v2)]\n",
    "            else:\n",
    "                vertices[v1] = Node(v1)\n",
    "                vertices[v1].outdegree += 1\n",
    "                edges[v1] = [Edge(v2)]\n",
    "            if v2 in edges.keys():\n",
    "                vertices[v2].indegree += 1\n",
    "            else:\n",
    "                vertices[v2] = Node(v2)\n",
    "                vertices[v2].indegree += 1\n",
    "                edges[v2] = []\n",
    "            i += 1\n",
    "\n",
    "    return (vertices, edges)\n",
    "\n",
    "def output_contigs(g):\n",
    "    \"\"\" Perform searching for Eulerian path in the graph to output genome assembly\"\"\"\n",
    "    V = g[0]\n",
    "    E = g[1]\n",
    "    # Pick starting node (the vertex with zero in degree)\n",
    "    start = list(V.keys())[0]\n",
    "    for k in V.keys():\n",
    "        if V[k].indegree < V[start].indegree:\n",
    "            start = k\n",
    "\n",
    "    contig = start\n",
    "    current = start\n",
    "    while len(E[current]) > 0:\n",
    "        # Pick the next node to be traversed (for now, at random)\n",
    "        next = E[current][0]\n",
    "        del E[current][0]\n",
    "        contig += next.label[-1]\n",
    "        current = next.label\n",
    "\n",
    "    return contig\n",
    "\n",
    "def print_graph(g):\n",
    "    \"\"\" Print the information in the graph to be (somewhat) presentable \"\"\"\n",
    "    V = g[0]\n",
    "    E = g[1]\n",
    "    for k in V.keys():\n",
    "        print(\"name: \", V[k].label, \". indegree: \", V[k].indegree, \". outdegree: \", V[k].outdegree)\n",
    "        print(\"Edges: \")\n",
    "        for e in E[k]:\n",
    "            print(e.label)\n",
    "        print()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
